packages/@uppy/provider-views/src/ProviderView/ProviderView.jsx

```diff
@@ -296,35 +232,89 @@ export default class ProviderView extends View {
     }
   }
 
-  async listAllFiles (path, files = null) {
-    files = files || [] // eslint-disable-line no-param-reassign
-    const res = await this.provider.list(path)
-    res.items.forEach((item) => {
-      if (!item.isFolder) {
-        files.push(item)
-      } else {
-        this.addFolder(item)
-      }
-    })
-    const moreFiles = res.nextPagePath
-    if (moreFiles) {
-      return this.listAllFiles(moreFiles, files)
+  async recursivelyListAllFiles (path, files = []) {
+    let curPath = path
+
+    // need to repeat the list call until there are no more pages
+    while (curPath) {
+      const res = await this.provider.list(curPath)
+
+      // limit concurrency (because what if we have a folder with 1000 folders inside!)
+      // todo this might still lead to a memory run-off
+      await pMap(res.items, async (item) => {
+        if (item.isFolder) {
+          // recursively call self for folder
+          await this.recursivelyListAllFiles(item.requestPath, files)
+        } else {
+          files.push(item)
+        }
+      }, { concurrency: 10 })
```

Comment:
It would indeed lead to a memory run-off, so let's avoid it if we can?
```suggestion
      for (const item of res.items)
        if (item.isFolder) {
          // recursively call self for folder
          yield* this.recursivelyListAllFiles(item.requestPath, files)
        } else {
          yield item
        }
      }
```

Reply:
cool! did you test this code? i'm not super confident with generators, but if you can vouch for it i'm good

Reply:
You can't commit my suggestion and expect it to work, there's more work that needs to happen to make code actually streamlined. But I think it's worth it so we can avoid running out of memory.

Reply:
You can't commit my suggestion and expect it to work, there's more work that needs to happen to make code actually streamlined. But I think it's worth it so we can avoid running out of memory.

Reply:
I just pushed https://github.com/transloadit/uppy/pull/4384/commits/50a0bece35b0218274b9f1f6446c7cc6f843ae95 – I didn't test it with Dropbox because I don't have a working Dropbox setup it seems, but it seems to work with Google Drive.